### PR TITLE
Refactoring of internal retry types

### PR DESCRIPTION
**Pull request by asmyasnikov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: #494

## What is the new behavior?

-
-
-

### CODE REVIEW

File: internal/xerrors/operation.go [hunk 1]
```diff
@@ -101,19 +100,19 @@ func IsOperationError(err error, codes ...Ydb.StatusIds_StatusCode) bool {
...
 		Ydb.StatusIds_BAD_SESSION,
 		Ydb.StatusIds_SESSION_BUSY:
-		return operation.NotFinished
+		return TypeRetryable
 	case Ydb.StatusIds_UNDETERMINED:
-		return operation.Undefined
+		return TypeConditionallyRetryable
 	default:
-		return operation.Finished
+		return TypeNonRetryable
```
Comment (rekby) [inline internal/xerrors/operation.go:115]: What about explicit add known unretriable operations as separate case and use default as fallback only.

File: internal/xerrors/type.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package xerrors
+
+import "fmt"
+
+// Type reports which error type
+type Type uint8
+
+const (
+	TypeNoError = Type(iota)
```
Comment (rekby) [inline internal/xerrors/type.go:9]: What about iota + 1 or add TypeUndefined as zero item?

for use explicit values only?

File: internal/xsql/conn.go [hunk 3]
```diff
@@ -139,7 +138,7 @@ func (c *conn) PrepareContext(ctx context.Context, query string) (_ driver.Stmt,
 
 func (c *conn) execContext(ctx context.Context, query string, args []driver.NamedValue) (_ driver.Result, err error) {
 	m := queryModeFromContext(ctx, c.defaultQueryMode)
-	onDone := trace.DatabaseSQLOnConnExec(c.trace, &ctx, query, m.String(), retry.IsIdempotent(ctx))
+	onDone := trace.DatabaseSQLOnConnExec(c.trace, &ctx, query, m.String(), xerrors.IsIdempotent(ctx))
```
Comment (rekby) [inline internal/xsql/conn.go:141]: What about move IsIdempotent, WithIdempotent from xerrors to xcontext?
```
xcontext.IsIdempotent(ctx)
```

Comment (asmyasnikov) [reply internal/xsql/conn.go:141]: ack

File: internal/xerrors/operation.go (hunk 1 above)
Comment (asmyasnikov) [reply internal/xerrors/operation.go:115]: ack

**Pull request merged**